### PR TITLE
feat(onboarding): use .editorconfig file to get indentation style

### LINKED ISSUE #4436: Use `.editorconfig` to get indentation style when renovate configuration doesn’t exist
## What Renovate type are you using?
The official GitHub App: https://github.com/marketplace/renovate

## Describe the bug
The default indentation is two spaces even when [`.editorconfig`](https://editorconfig.org/) is configured to use tabs, because tabs are the superior indentation codepoint (<https://github.com/prettier/prettier/issues/7475>).

## To Reproduce

1.  Create repository with `.editorconfig` set to:

    ```properties
    root = true

    [*]
    # Use tabs, because tabs are the superior indentation codepoint
    # https://github.com/prettier/prettier/issues/7475
    indent_style = tab
    ```
2.  Let Renovate create the onboarding PR.

## Expected behavior
`renovate.json` is created with:
```json
{
	"extends": [
		"config:base"
	]
}
```

## Actual behavior
`renovate.json` is created with:
```json
{
  "extends": [
    "config:base"
  ]
}
```

### PR DESCRIPTION
**Pull request by pret-a-porter**
## Changes:

1. Introduced `EditorConfig` class to read `.editorconfig` file and returns code format for specific file
2. Introduce `JsonWriter` class to incapsulate code formatting for json files

## Context:

Closes #4436 

## Documentation

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Review by tPlouis99 — APPROVED** [review]

File: lib/util/json-writer/editor-config.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import globToRegExp from 'glob-to-regexp';
+import { logger } from '../../logger/index';
+import { localPathExists, readLocalFile } from '../fs';
+import type { CodeFormat } from './code-format';
+import { IndentationType } from './indentation-type';
+
+export class EditorConfig {
+  private readonly formats: Map<RegExp, KnownProps> = new Map();
+
+  private static instance?: EditorConfig;
```
Comment (viceice) [inline lib/util/json-writer/editor-config.ts:12]: Please use memory cache instead, so we can auto cleanup after repo processing

https://github.com/renovatebot/renovate/blob/095f2f8240ca90d3dc61016cb97b84d8e853c9b8/lib/util/cache/memory/index.ts#L15

File: lib/util/json-writer/editor-config.ts [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+  private constructor(parseResult: ParseStringResult = []) {
+    for (const [sectionName, sectionBody] of parseResult) {
+      const regExp = globToRegExp(sectionName ?? '*');
```
Comment (viceice) [inline lib/util/json-writer/editor-config.ts:16]: Can't we use the glob, as we already use a glob lib ( `minimatch`) .

File: lib/util/json-writer/json-writer.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
+import { IndentationType } from './indentation-type';
+
+export class JSONWriter {
+  private readonly indentationType: IndentationType;
+
+  private readonly indentationSize: number;
+
+  constructor(type = IndentationType.Space, size = 2) {
```
Comment (viceice) [inline lib/util/json-writer/json-writer.ts:8]: Can't we simply pass the `CodeFormat` object?

File: lib/util/json-writer/editor-config.ts [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    return EditorConfig.instance;
+  }
+
+  public static reset(): void {
+    delete this.instance;
+  }
+
+  public getCodeFormat(fileName: string): CodeFormat {
```
Comment (viceice) [inline lib/util/json-writer/editor-config.ts:41]: I think we should use `EditorConfig.parseFromFiles` api for reading config options

https://www.npmjs.com/package/editorconfig#parsefromfilesfilepath-configs-options

**Review by viceice — CHANGES_REQUESTED** [review]
@rarkins We also have unused `detect-indent` dependency. 🤔

File: lib/util/json-writer/editor-config.ts (hunk 1 above)
Comment (pret-a-porter) [reply lib/util/json-writer/editor-config.ts:12]: In one nice day we will receive collision of names with this approach

File: lib/util/json-writer/editor-config.ts (hunk 4 above)
Comment (pret-a-porter) [reply lib/util/json-writer/editor-config.ts]: Nice idea, eventually I am using `parseSync`

**Comment by rarkins** [comment]
Needs deconflicting

File: lib/util/json-writer/editor-config.spec.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
+import mock from 'mock-fs';
```
Comment (viceice) [inline lib/util/json-writer/editor-config.spec.ts:1]: I'm unsure about this new dependency. @JamieMagee @zharinov WDYT?

**Comment by pret-a-porter** [comment]
Deconflicted

File: lib/util/json-writer/editor-config.spec.ts (hunk 5 above)
Comment (JamieMagee) [reply lib/util/json-writer/editor-config.spec.ts:1]: Yeah, we normally include a `__fixtures__` directory, and put test fixtures in there. Separates out the tests from and sample data.

File: lib/util/json-writer/editor-config.ts [hunk 6]
```diff
@@ -0,0 +1,42 @@
+import { KnownProps, parseSync } from 'editorconfig';
+import type { CodeFormat } from './code-format';
+import { IndentationType } from './indentation-type';
+
+export class EditorConfig {
+  public static getCodeFormat(fileName: string): CodeFormat {
+    const knownProps = parseSync(fileName);
```
Comment (rarkins) [inline lib/util/json-writer/editor-config.ts:7]: We don't like every using sync methods, can it be switched to async?

Comment (rarkins) [reply lib/util/json-writer/editor-config.ts:7]: And how far "up" does this method go - e.g. would it find an `.editorconfig` in the root directory if one existed?

**Review by rarkins — CHANGES_REQUESTED** [review]

File: lib/util/json-writer/editor-config.ts (hunk 6 above)
Comment (pret-a-porter) [reply lib/util/json-writer/editor-config.ts:7]: https://github.com/editorconfig/editorconfig-core-js#parsefilepath-options yeah, it should find and `.editorconfig` file in the root directory

Comment (pret-a-porter) [reply lib/util/json-writer/editor-config.ts:7]: I have tried rewrite to async await, but by some reason unit tests fails, therefore I propose to go on with `parseSync`

Comment (viceice) [reply lib/util/json-writer/editor-config.ts:7]: Please use async and push changes, we can then have a look why it fails.

Comment (pret-a-porter) [reply lib/util/json-writer/editor-config.ts]: Done. Seems like content for .editorconfig file does not mock properly by some reason

**Comment by rarkins** [comment]
@pret-a-porter please run it against some test repos on github.com and then link us to the results

**Comment by rarkins** [comment]
@pret-a-porter have you run against real repos, and can you share the results?

**Review by rarkins — DISMISSED** [review]

File: lib/util/json-writer/editor-config.spec.ts [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  });
+
+  afterEach(() => {
+    mockFs.restore();
+  });
+
+  afterAll(() => {
+    setGlobalConfig();
+  });
+
```
Comment (viceice) [inline lib/util/json-writer/editor-config.spec.ts:29]:
```suggestion
```

I don't think we need this.

File: lib/workers/repository/onboarding/branch/config.spec.ts [hunk 8]
```diff
@@ -9,16 +10,28 @@ const mockedPresets = presets as jest.Mocked<typeof presets>;
...
     jest.clearAllMocks();
     config = getConfig();
     config.platform = 'github';
     config.repository = 'some/repo';
   });
+
+  afterAll(() => {
+    setGlobalConfig();
+  });
+
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/config.spec.ts:30]:
```suggestion
```

same as in other file

File: lib/workers/repository/onboarding/branch/rebase.spec.ts [hunk 9]
```diff
@@ -1,9 +1,20 @@
...
+  beforeAll(() => {
+    setGlobalConfig({
+      localDir: '',
+    });
+  });
+
+  afterAll(() => {
+    setGlobalConfig();
+  });
+
```
Comment (viceice) [inline lib/workers/repository/onboarding/branch/rebase.spec.ts:17]:
```suggestion
```

**Review by viceice — DISMISSED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**